lib/internal/fetch/headers.js

```diff
@@ -0,0 +1,213 @@
+        }
+        headers.append(header[0], header[1]);
+      }
+    } else if (typeof object[0] === 'string' || Buffer.isBuffer(object[0])) {
+      // Flat array of strings or Buffers
+      if (object.length % 2 !== 0) {
+        throw new ERR_INVALID_ARG_VALUE('init', object,
+                                        'is not even in length');
+      }
+      for (let i = 0; i < object.length; i += 2) {
+        headers.append(
+          object[i].toString('utf-8'),
+          object[i + 1].toString('utf-8')
+        );
+      }
+    } else {
+      // All other array based entries
+      throw new ERR_INVALID_ARG_VALUE(
+        'init',
+        object,
+        'is not a valid array entry'
+      );
+    }
+  } else if (!isBoxedPrimitive(object)) {
+    // Object of key/value entries
+    for (const { 0: name, 1: value } of ObjectEntries(object)) {
+      headers.append(name, value);
+    }
+  }
+}
+
+class Headers {
+  constructor(init) {
+    this[kHeadersList] = [];
+
+    if (init &&
+      validateObject(init, 'init', { allowArray: true }) === undefined) {
+      fill(this, init);
+    }
+  }
+
+  append(name, value) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+    const normalizedValue = normalizeAndValidateHeaderValue(name, value);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    if (this[kHeadersList][index] === normalizedName) {
+      this[kHeadersList][index + 1] += `, ${normalizedValue}`;
+    } else {
+      this[kHeadersList].splice(index, 0, normalizedName, normalizedValue);
+    }
+  }
+
+  delete(name) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    if (this[kHeadersList][index] === normalizedName) {
+      this[kHeadersList].splice(index, 2);
+    }
+  }
+
+  get(name) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    if (this[kHeadersList][index] === normalizedName) {
+      return this[kHeadersList][index + 1];
+    }
+
+    return null;
+  }
+
+  has(name) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    return this[kHeadersList][index] === normalizedName;
+  }
+
+  set(name, value) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+    const normalizedValue = normalizeAndValidateHeaderValue(name, value);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+    if (this[kHeadersList][index] === normalizedName) {
+      this[kHeadersList][index + 1] = normalizedValue;
+    } else {
+      this[kHeadersList].splice(index, 0, normalizedName, normalizedValue);
+    }
+  }
+
+  * keys() {
+    for (const header of this) {
+      yield header[0];
+    }
+  }
+
+  * values() {
+    for (const header of this) {
+      yield header[1];
+    }
```

Comment:
Wouldn't it be more efficient to avoid the array creation?
```suggestion
    for (let index = 0; index < this[kHeadersList].length; index += 2) {
      yield this[kHeadersList][index];
    }
  }

  * values() {
    for (let index = 1; index < this[kHeadersList].length; index += 2) {
      yield this[kHeadersList][index];
    }
```